muxfs.1
.\" Copyright (c) 2022 Stephen D. Adams <stephen@sdadams.org>
.\" 
.\" Permission to use, copy, modify, and distribute this software for any
.\" purpose with or without fee is hereby granted, provided that the above
.\" copyright notice and this permission notice appear in all copies.
.\" 
.\" THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
.\" WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
.\" MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
.\" ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
.\" WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
.\" ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
.\" OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
.\"
.Dd $Mdocdate: July 8 2022 $
.Dt MUXFS 1
.Os
.Sh NAME
.Nm muxfs
.Nd the Multiplexed File System
.Sh SYNOPSIS
.Nm muxfs
.Cm format
.Op Fl a Ar checksum_algorithm
.Ar directory ...
.Nm muxfs
.Cm mount
.Op Fl f
.Ar mount_point
.Ar directory ...
.Nm muxfs
.Cm audit
.Ar directory ...
.Nm muxfs
.Cm heal
.Ar directory ...
.Nm muxfs
.Cm sync
.Ar destination
.Ar source ...
.Nm muxfs
.Cm version
.Sh DESCRIPTION
.Nm
is a mirroring, checksumming, and self-healing filesystem layer for OpenBSD.
It mirrors a filesystem tree across a series of directories, and employs
checksum databases to validate and restore files automatically.
.Pp
In addition to the file content, metadata is checksummed, including: owner,
group, mode, (external) inode number, and for regular files the file size.  For
directories the checksummed content includes the filenames and corresponding
checksums of the files in the directory.  For symbolic links the checksummed
content is the path output by
.Xr readlink 2 .
.Sh COMMANDS
During an
.Cm audit ,
.Cm heal ,
or
.Cm sync
the
.Nm
array must not be mounted.
.Bl -tag -width Ds
.It Cm format Oo Fl a Ar checksum_algorithm Oc Ar directory ...
Format the directories as part of a new
.Nm
array.  The
.Ar checksum_algorithm
may be
.Cm crc32 ,
.Cm md5 ,
or
.Cm sha1 ,
and defaults to
.Cm md5 .
The directories must be empty and non-overlapping.
.It Cm mount Oo Fl f Oc Ar mount_point Ar directory ...
Mount the directories as a redundant array at
.Ar mount_point .
If
.Fl f
is specified then
.Nm
remains in the foreground.
.It Cm audit Ar directory ...
If the directories are out-of-sync write a warning to the standard error and exit, otherwise for each
.Ar directory
read and check the checksums of all filesystem nodes in the
.Ar directory ,
and write to the standard output the paths of the filesystem nodes that do not
match.
.It Cm heal Ar directory ...
Perform an
.Cm audit
and attempt to restore the filesystem nodes found to be corrupted using the
corresponding copies in the other directories.
.It Cm sync Ar destination Ar source ...
Restore the directory
.Ar destination
to match the state of the
.Ar source
directories.
.It Cm version
Display the program version and exit.
.Sh EXIT STATUS
.Nm
exits with one of the following exit status values: 
.Pp
.Bl -tag -width Ds -offset indent -compact
.It 0
success
.It 1
incorrect usage
.It 255
error
.Sh EXAMPLES
Create a new
.Nm
array from the filesystems mounted at
.Pa /mnt/a
and
.Pa /mnt/b
using the
.Cm sha1
checksum algorithm:
.Bd -literal -offset indent
# muxfs format -a sha1 /mnt/a /mnt/b
.Ed
.Pp
Mount the
.Nm
array consisting of directories
.Pa /mnt/a
and
.Pa /mnt/b
to the mount-point
.Pa /mnt/mp :
.Bd -literal -offset indent
# muxfs mount /mnt/mp /mnt/a /mnt/b
.Ed
.Pp
Unmount the
.Nm
array:
.Bd -literal -offset indent
# umount /mnt/mp
.Ed
.Pp
To perform an
.Cm audit
or
.Cm heal
operation the array must not be mounted.
.Pp
Audit the
.Nm
array:
.Bd -literal -offset indent
# muxfs audit /mnt/a /mnt/b
.Ed
.Pp
Heal the
.Nm
array:
.Bd -literal -offset indent
# muxfs heal /mnt/a /mnt/b
.Ed
.Pp
Re-sync corrupted
.Pa /mnt/b
to state of
.Pa /mnt/a
after loss of power:
.Bd -literal -offset indent
# muxfs sync /mnt/b /mnt/a
.Ed
.Pp
Create or update a backup of the array to
.Pa /mnt/c ,
the mount-point of a filesystem on a detachable disk:
.Bd -literal -offset indent
# muxfs sync /mnt/c /mnt/a /mnt/b
.Ed
.Sh CAVEATS
.Nm
supports regular files, directories, and symbolic links, and does not support
special files, device nodes, file locking, or hard-linking.  Timestamps are not
included in the checksummed metadata.
.Pp
Filesystem nodes named
.Pa .muxfs
are reserved for use by
.Nm .
.Pp
.Nm
relies on the uniqueness of inode numbers within each directory to perform
correctly, therefore the mounting of filesystems within array directories is
not supported, with the exception of the
.Pa .muxfs
directories.
.Pp
To mitigate hard drive head thrashing the top-level
.Pa .muxfs
directories may be moved to solid state storage then mounted back to their
original locations.
.Pp
.Nm
relies on writing sequentially to multiple directories to mitigate corruption
during unexpected of loss of power.  After such an event the administrator must
.Cm sync
the array from one or more of the uncorrupted directories before mounting it.
Due to this it is not recommended to use muxfs with only one directory in the
array.
.Pp
For privacy between users
.Xr syslog 3
messages are sent to the
.Sq user
facility.  To receive these messages an entry must be set up in
.Xr syslog.conf 5 .
Ensure that the corresponding log file is not readable by other.  The log
messages include notifications of files found to be corrupted and the success
or failure of file restoration attempts.
.Pp
The size of the
.Nm
database is proportional to the value of the largest (internal) inode number
encountered.  To minimize this dedicate whole filesystems to the
.Nm
array.
.Sh AUTHORS
.An Stephen D. Adams Aq Mt stephen@sdadams.org


CHANGES
0.3-current (2022-07-14):
* Replaced the multiple binaries with a single binary with sub-commands.
* Renamed "newfs_muxfs" to "muxfs format".
* Renamed "fsck_muxfs" to "muxfs heal"
* Renamed "fsck_muxfs -a" to "muxfs audit".
* Added the command "sync" to allow for drive replacement.
* Replaced the array of UUIDs in the conf with single UUID for the whole-array.
* Added a "restoring" field to the state.db file.
* Added manual page.
* Updated README and GLOSSARY.
* Added checking of sequence numbers before mount, audit, and heal.
* Replaced first-come-first-serve statfs() with aggregated statfs().
* Bug fixes.

0.2-current (2022-06-28):
* Optimization for large files.
* Added fsck_muxfs.
* Removed the redundant 'type' from the metadata checksum.
* Added the 'size' to the metadata checksum for regular files.
* Removed 'sign' from the configuration file.
* Updated README and GLOSSARY.
* Bug fixes.

0.1-current (2022-06-22):
* Test suite.
* Access from other users.
* Dynamic stack allocator.
* Bug fixes.

0.0-current (2022-06-10):
* Initial publication.


muxfs.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "ds.h"
#include "muxfs.h"

static void
usage(void)
{
	dprintf(2, "usage: muxfs format [-a checksum_algorithm] "
	    "directory ...\n");
	dprintf(2, "       muxfs mount [-f] mount_point directory ...\n");
	dprintf(2, "       muxfs audit directory ...\n");
	dprintf(2, "       muxfs heal directory ...\n");
	dprintf(2, "       muxfs sync destination source ...\n");
	dprintf(2, "       muxfs version\n");
}

int
main(int argc, char *argv[])
{
	const char *cmd;

	if (muxfs_state_syslog_init())
		exit(-1);
	if (muxfs_dsinit())
		exit(-1);

	if (argc < 2) {
		usage();
		exit(-1);
	}
	cmd = argv[1];

	if (strcmp(cmd, "format") == 0)
		return muxfs_format_main(argc, argv);
	else if (strcmp(cmd, "mount") == 0)
		return muxfs_mount_main(argc, argv);
	else if (strcmp(cmd, "audit") == 0)
		return muxfs_scan_main(MUXFS_SCAN_AUDIT, argc, argv);
	else if (strcmp(cmd, "heal") == 0)
		return muxfs_scan_main(MUXFS_SCAN_HEAL, argc, argv);
	else if (strcmp(cmd, "sync") == 0)
		return muxfs_sync_main(argc, argv);
	else if (strcmp(cmd, "version") == 0) {
		muxfs_version_print();
		exit(0);
	}

	usage();
	exit(-1);
}
